### ISSUE TITLE
about how to generate test result

### ISSUE THREAD

**Issue by SupOne11** [ZikangZhou/HiVT#42]
I write a test code for generate test result, but the FDE(k=6) of the pretrained model show a 0.1 decrease compare to the HIVT128 on the leaderboard. Could you show me your test.py.? Thank you.

**Comment by LL-C8** [comment]
@SupOne11 Could you please show me your complete test.py.? Thank you！

**Comment by SupOne11** [comment]
from argparse import ArgumentParser

import pytorch_lightning as pl
import torch
from torch_geometric.data import DataLoader

from datasets import ArgoverseV1Dataset
from models.hivt import HiVT

from tqdm import tqdm
import torch.nn.functional as F
from utils import plot_single_vehicle, rotate_and_translate
import matplotlib.pyplot as plt
from argoverse.map_representation.map_api import ArgoverseMap
from argoverse.evaluation.eval_forecasting import compute_forecasting_metrics


# --root datasets/to/dataset_root/ --batch_size 32 --ckpt_path lightning_logs/version_20/checkpoints/epoch=63-step=411903.ckpt
if __name__ == '__main__':
    pl.seed_everything(2022)

    parser = ArgumentParser()
    parser.add_argument('--root', type=str, required=True)
    parser.add_argument('--batch_size', type=int, default=32)
    parser.add_argument('--num_workers', type=int, default=16)
    parser.add_argument('--pin_memory', type=bool, default=True)
    parser.add_argument('--persistent_workers', type=bool, default=True)
    parser.add_argument('--gpus', type=int, default=1)
    parser.add_argument('--plot', default=True)
    parser.add_argument('--ckpt_path', type=str, required=True)
    args = parser.parse_args()

    model = HiVT.load_from_checkpoint(checkpoint_path=args.ckpt_path, parallel=True)
    model.to('cuda:0')
    test_dataset = ArgoverseV1Dataset(root=args.root, split='test', local_radius=model.hparams.local_radius)
    dataloader = DataLoader(test_dataset, batch_size=args.batch_size, shuffle=False, num_workers=args.num_workers,
                            pin_memory=args.pin_memory, persistent_workers=args.persistent_workers)

    preds, prob = {}, {}
    gts, cities = {}, {}
    for ii, data in tqdm(enumerate(dataloader)):
        with torch.no_grad():
            data.to('cuda:0')
            output, pi = model(data)
            index_ag = data['agent_index']
            origin_ag = data['positions'][index_ag, 19, :]
            result_all = output.transpose(1, 0)  # [N, mod, 30, 2]
            result_ag = result_all[index_ag, :, :, 0:2]  # [agent, mod, 30, 2]只保留agent的预测结果
            result_av = rotate_and_translate(result_ag, -data['rotate_angles'][index_ag], origin_ag)  # 将预测还原到av坐标系
            result = rotate_and_translate(result_av, -data['theta'], data['origin'])  # 将预测还原到city坐标系
            result_list = [x.detach().cpu().numpy() for x in result]  # 预测结果存到列表中
            cls = F.softmax(pi[index_ag], dim=-1)
            cls_list = [x.detach().cpu().numpy() for x in cls]
            for i in range(data['seq_id'].size(0)):
                seq_id = data['seq_id'][i].cpu().item()
                preds[seq_id] = result_list[i]
                prob[seq_id] = cls_list[i].squeeze()
                cities[seq_id] = data['city'][i]

    from argoverse.evaluation.competition_util import generate_forecasting_h5

    generate_forecasting_h5(data=preds, output_path="submit.h5", probabilities=prob)

**Comment by SupOne11** [comment]
from argparse import ArgumentParser
import pytorch_lightning as pl
import torch
from torch_geometric.data import DataLoader
from datasets import ArgoverseV1Dataset
from models.hivt import HiVT
from tqdm import tqdm
import torch.nn.functional as F
from utils import plot_single_vehicle, rotate_and_translate
import matplotlib.pyplot as plt
from argoverse.map_representation.map_api import ArgoverseMap
from argoverse.evaluation.eval_forecasting import compute_forecasting_metrics


if __name__ == '__main__':
    pl.seed_everything(2022)

    parser = ArgumentParser()
    parser.add_argument('--root', type=str, required=True)
    parser.add_argument('--batch_size', type=int, default=32)
    parser.add_argument('--num_workers', type=int, default=16)
    parser.add_argument('--pin_memory', type=bool, default=True)
    parser.add_argument('--persistent_workers', type=bool, default=True)
    parser.add_argument('--gpus', type=int, default=1)
    parser.add_argument('--plot', default=True)
    parser.add_argument('--ckpt_path', type=str, required=True)
    args = parser.parse_args()

    model = HiVT.load_from_checkpoint(checkpoint_path=args.ckpt_path, parallel=True)
    model.to('cuda:0')
    test_dataset = ArgoverseV1Dataset(root=args.root, split='test', local_radius=model.hparams.local_radius)
    dataloader = DataLoader(test_dataset, batch_size=args.batch_size, shuffle=False, num_workers=args.num_workers,
                            pin_memory=args.pin_memory, persistent_workers=args.persistent_workers)

    preds, prob = {}, {}
    gts, cities = {}, {}
    for ii, data in tqdm(enumerate(dataloader)):
        with torch.no_grad():
            data.to('cuda:0')
            output, pi = model(data)
            index_ag = data['agent_index']
            origin_ag = data['positions'][index_ag, 19, :]
            result_all = output.transpose(1, 0)  # [N, mod, 30, 2]
            result_ag = result_all[index_ag, :, :, 0:2]  # [agent, mod, 30, 2]只保留agent的预测结果
            result_av = rotate_and_translate(result_ag, -data['rotate_angles'][index_ag], origin_ag)  # 将预测还原到av坐标系
            result = rotate_and_translate(result_av, -data['theta'], data['origin'])  # 将预测还原到city坐标系
            result_list = [x.detach().cpu().numpy() for x in result]  # 预测结果存到列表中
            cls = F.softmax(pi[index_ag], dim=-1)
            cls_list = [x.detach().cpu().numpy() for x in cls]
            for i in range(data['seq_id'].size(0)):
                seq_id = data['seq_id'][i].cpu().item()
                preds[seq_id] = result_list[i]
                prob[seq_id] = cls_list[i].squeeze()
                cities[seq_id] = data['city'][i]

    from argoverse.evaluation.competition_util import generate_forecasting_h5

    generate_forecasting_h5(data=preds, output_path="submit.h5", probabilities=prob)

**Comment by SupOne11** [comment]
@LCY-coder 
from argparse import ArgumentParser
import pytorch_lightning as pl
import torch
from torch_geometric.data import DataLoader
from datasets import ArgoverseV1Dataset
from models.hivt import HiVT
from tqdm import tqdm
import torch.nn.functional as F
from utils import plot_single_vehicle, rotate_and_translate
import matplotlib.pyplot as plt
from argoverse.map_representation.map_api import ArgoverseMap
from argoverse.evaluation.eval_forecasting import compute_forecasting_metrics


if __name__ == '__main__':
    pl.seed_everything(2022)

    parser = ArgumentParser()
    parser.add_argument('--root', type=str, required=True)
    parser.add_argument('--batch_size', type=int, default=32)
    parser.add_argument('--num_workers', type=int, default=16)
    parser.add_argument('--pin_memory', type=bool, default=True)
    parser.add_argument('--persistent_workers', type=bool, default=True)
    parser.add_argument('--gpus', type=int, default=1)
    parser.add_argument('--plot', default=True)
    parser.add_argument('--ckpt_path', type=str, required=True)
    args = parser.parse_args()

    model = HiVT.load_from_checkpoint(checkpoint_path=args.ckpt_path, parallel=True)
    model.to('cuda:0')
    test_dataset = ArgoverseV1Dataset(root=args.root, split='test', local_radius=model.hparams.local_radius)
    dataloader = DataLoader(test_dataset, batch_size=args.batch_size, shuffle=False, num_workers=args.num_workers,
                            pin_memory=args.pin_memory, persistent_workers=args.persistent_workers)

    preds, prob = {}, {}
    gts, cities = {}, {}
    for ii, data in tqdm(enumerate(dataloader)):
        with torch.no_grad():
            data.to('cuda:0')
            output, pi = model(data)
            index_ag = data['agent_index']
            origin_ag = data['positions'][index_ag, 19, :]
            result_all = output.transpose(1, 0)  # [N, mod, 30, 2]
            result_ag = result_all[index_ag, :, :, 0:2]  # [agent, mod, 30, 2]
            result_av = rotate_and_translate(result_ag, -data['rotate_angles'][index_ag], origin_ag)  #
            result = rotate_and_translate(result_av, -data['theta'], data['origin'])  
            result_list = [x.detach().cpu().numpy() for x in result]  
            cls = F.softmax(pi[index_ag], dim=-1)
            cls_list = [x.detach().cpu().numpy() for x in cls]
            for i in range(data['seq_id'].size(0)):
                seq_id = data['seq_id'][i].cpu().item()
                preds[seq_id] = result_list[i]
                prob[seq_id] = cls_list[i].squeeze()
                cities[seq_id] = data['city'][i]

    from argoverse.evaluation.competition_util import generate_forecasting_h5

    generate_forecasting_h5(data=preds, output_path="submit.h5", probabilities=prob)

**Comment by SupOne11** [comment]
def rotate_and_translate(x, theta, origin):
    if x.dim() == 3:
        x = x.unsqueeze(1)
    batch = x.size(0)
    mod = x.size(1)
    pt_num = x.size(2)
    sin_vals = torch.sin(theta)
    cos_vals = torch.cos(theta)
    rotate_mat = torch.empty(batch, 2, 2, device='cuda:0')
    rotate = torch.empty(mod, batch, pt_num, 2, device='cuda:0')
    rotate_mat[:, 0, 0] = cos_vals
    rotate_mat[:, 0, 1] = -sin_vals
    rotate_mat[:, 1, 0] = sin_vals
    rotate_mat[:, 1, 1] = cos_vals
    x = x.transpose(1, 0)
    for i in range(mod):
        rotate[i] = torch.bmm(x[i], rotate_mat)
    return rotate.transpose(1, 0) + origin.view(batch, 1, 1, 2)

**Comment by wangqc1001** [comment]
Could you please share with me how to use the trained model and visualize the predictions while reading the test set?

**Comment by Zhangxx1218** [comment]
May I ask how it gets the model results ah, he doesn't use to the test set but the validation set When running eval.py, I don't reproduce the paper metrics. Did you reproduce it?

**Comment by HHHzp** [comment]
I also wrote a test. py, but tested it with a pre trained model and the results were worse than those in the paper.
